test/mailer_notifier_test.exs

```diff
@@ -34,330 +34,354 @@ defmodule MailerNotifierTest do
     end
   end
 
-  describe "Bamboo" do
-    defmodule TestRouter do
-      use Phoenix.Router
-      import Phoenix.Controller
-
-      use BoomNotifier,
-        notifier: BoomNotifier.MailNotifier.Bamboo,
-        options: [
-          mailer: Support.FakeMailer,
-          from: "me@example.com",
-          to: self(),
-          subject: "BOOM error caught"
-        ],
-        custom_data: [:assigns, :logger]
-
-      pipeline :browser do
-        plug(:accepts, ["html"])
-        plug(:save_custom_data)
+  # All mail notifiers should behave the same, so define the tests across
+  # the set of notifiers.
+  #
+  # Note: Because `use` is macro, we can't inject `mail_notifier_module`
+  #       as the `notifier` value directly, because it wont be resolved
+  #       in the correct context. Instead we can define a module attribute
+  #       which gets replaced at compile time with correct atom.
+  #
+  #       We also disable module-redefinition warning just in this context,
+  #       since we are redefining the router modules for each notifier.
+
+  notifiers = [
+    {"Bamboo", BoomNotifier.MailNotifier.Bamboo},
+    {"Swoosh", BoomNotifier.MailNotifier.Swoosh}
+  ]
+
+  for {name, mail_notifier_module} <- notifiers do
+    describe name do
+      Code.put_compiler_option(:ignore_module_conflict, true)
+
+      defmodule TestRouter do
+        use Phoenix.Router
+        import Phoenix.Controller
+
+        @mail_notifier_module mail_notifier_module
+        use BoomNotifier,
+          notifier: @mail_notifier_module,
```

Comment:
We can't do

```
use BoomNotifier,
   notifier: mail_notifier_module,
   ...
```

Because the `__using__` macro doesn't expand the value and just tries to use `mail_notifier_module` as an actual variable. Using the attribute is the cleanest way I could find. Some posts say to use `unquote` here, but that doesn't work in our case because we're already unquoting in the macro and you can't unquote inside an unquote.